core/genesis.go

```diff
@@ -290,177 +371,121 @@ type ChainOverrides struct {
-			}
-			if overrides != nil && overrides.OverrideOptimismFjord != nil {
-				config.FjordTime = overrides.OverrideOptimismFjord
-			}
-			if overrides != nil && overrides.OverrideOptimismGranite != nil {
-				config.GraniteTime = overrides.OverrideOptimismGranite
-			}
-			if overrides != nil && overrides.OverrideOptimismHolocene != nil {
-				config.HoloceneTime = overrides.OverrideOptimismHolocene
-			}
-			if overrides != nil && overrides.OverrideOptimismInterop != nil {
-				config.InteropTime = overrides.OverrideOptimismInterop
-			}
-		}
+		return nil, common.Hash{}, nil, errGenesisNoConfig
 	}
-	// Just commit the new block if there is no stored genesis block.
-	stored := rawdb.ReadCanonicalHash(db, 0)
-	if (stored == common.Hash{}) {
+	// Commit the genesis if the database is empty
+	ghash := rawdb.ReadCanonicalHash(db, 0)
+	if (ghash == common.Hash{}) {
 		if genesis == nil {
 			log.Info("Writing default main-net genesis block")
 			genesis = DefaultGenesisBlock()
 		} else {
 			log.Info("Writing custom genesis block")
 		}
+		chainCfg, err := overrides.apply(genesis.Config)
+		if err != nil {
+			return nil, common.Hash{}, nil, err
+		}
+		genesis.Config = chainCfg
 
-		applyOverrides(genesis.Config)
 		block, err := genesis.Commit(db, triedb)
 		if err != nil {
-			return genesis.Config, common.Hash{}, err
+			return nil, common.Hash{}, nil, err
 		}
-		return genesis.Config, block.Hash(), nil
+		return chainCfg, block.Hash(), nil, nil
 	}
-	// The genesis block is present(perhaps in ancient database) while the
-	// state database is not initialized yet. It can happen that the node
-	// is initialized with an external ancient store. Commit genesis state
-	// in this case.
+
+	// TODO: Reinstate transitioned network check. Not sure where exactly to put it after this refactor.
 	// If the bedrock block is not 0, that implies that the network was migrated at the bedrock block.
 	// In this case the genesis state may not be in the state database (e.g. op-geth is performing a snap
 	// sync without an existing datadir) & even if it were, would not be useful as op-geth is not able to
 	// execute the pre-bedrock STF.
-	header := rawdb.ReadHeader(db, stored, 0)
-	transitionedNetwork := genesis != nil && genesis.Config != nil && genesis.Config.BedrockBlock != nil && genesis.Config.BedrockBlock.Uint64() != 0
-	if header.Root != types.EmptyRootHash && !triedb.Initialized(header.Root) && !transitionedNetwork {
+	// transitionedNetwork := genesis != nil && genesis.Config != nil && genesis.Config.BedrockBlock != nil && genesis.Config.BedrockBlock.Uint64() != 0
+
+	// Commit the genesis if the genesis block exists in the ancient database
+	// but the key-value database is empty without initializing the genesis
+	// fields. This scenario can occur when the node is created from scratch
+	// with an existing ancient store.
+	storedCfg := rawdb.ReadChainConfig(db, ghash)
+	if storedCfg == nil {
```

Comment:
Maybe
```suggestion
	if storedCfg == nil && !transitionedNetwork {
```
? I'm not sure.